### PR TITLE
[LPT] ConvolutionBackpropData support

### PR DESCRIPTION
**Pull request by vzinovie**
issue: 45623
validation:

- accuracy:
  - branch: developer-services/job/accuracy/1948/
  - reference: developer-services/job/accuracy/1949/
- performance:
  - branch: developer-services/job/performance/853/
  - reference: developer-services/job/performance/854/

### CODE REVIEW

File: inference-engine/src/low_precision_transformations/src/network_helper.cpp [hunk 1]
```diff
@@ -1525,6 +1549,50 @@ std::shared_ptr<Node> NetworkHelper::toScalarIfPossible(std::shared_ptr<Node> no
...
+        if (!dataPrecision.hasZeroPoint) {
+            return true;
+        }
+        min = dataPrecision.min - 0.5f;
+        max = dataPrecision.max + 0.5f;
+        const auto quantizationDetails = QuantizationDetails::getDetails(as_type_ptr<opset1::FakeQuantize>(node));
+        for (size_t i = 0; i < quantizationDetails.outputIntervalsCount; ++i) {
+            const float shift =
+                    (dataPrecision.min * quantizationDetails.outputHighValues[i] - dataPrecision.max * quantizationDetails.outputLowValues[i]) /
+                    (quantizationDetails.outputHighValues[i] - quantizationDetails.outputLowValues[i]);
```
Comment (eshoguli) [inline inference-engine/src/low_precision_transformations/src/network_helper.cpp:1587]: should we check `quantizationDetails.outputHighValues[i]` and `quantizationDetails.outputLowValues[i]` equality and specially handle that?

File: inference-engine/src/low_precision_transformations/src/weightable_layer_transformation.cpp [hunk 2]
```diff
@@ -115,10 +115,18 @@ bool WeightableLayerTransformation::canBeTransformed(const TransformationContext
...
+                (shape_size(constOutputShape) != constOutputShape[0]) ||
+                ((constOutputShape[0] != 1ul) && (fqFromWeights->get_output_shape(0)[0] != constOutputShape[0]))) {
+                return false;
+            }
+        } else {
+            if ( // Check if all dimensions of scale except the second one (which is O-Output channels dimension) are all ones
+                (shape_size(constOutputShape) != constOutputShape[1]) ||
+                ((constOutputShape[1] != 1ul) && (fqFromWeights->get_output_shape(0)[1] != constOutputShape[1]))) {
+                return false;
+            }
```
Comment (eshoguli) [inline inference-engine/src/low_precision_transformations/src/weightable_layer_transformation.cpp:129]: is it copy/paste? can we specify `outChannelsShapeIndex` and reuse?
```cpp
const size_t outChannelsShapeIndex = is_type<opset1::ConvolutionBackpropData>(layer) ? 1ul : 0ul;
if ( // Check if all dimensions of scale except the second one (which is O-Output channels dimension) are all ones
    (shape_size(constOutputShape) != constOutputShape[outChannelsShapeIndex]) ||
    ((constOutputShape[outChannelsShapeIndex] != 1ul) && (fqFromWeights->get_output_shape(0)[outChannelsShapeIndex] != constOutputShape[outChannelsShapeIndex]))) {
    return false;
}
```

File: inference-engine/src/transformations/include/ngraph_ops/deconvolution_ie.hpp [hunk 3]
```diff
@@ -29,6 +29,7 @@ class TRANSFORMATIONS_API DeconvolutionIE : public Op {
                     const Strides& dilations,
                     const CoordinateDiff& pads_begin,
                     const CoordinateDiff& pads_end,
+                    const element::Type output_type,
```
Comment (eshoguli) [inline inference-engine/src/transformations/include/ngraph_ops/deconvolution_ie.hpp:32]: we don't need it

File: inference-engine/src/transformations/src/ngraph_ops/deconvolution_ie.cpp [hunk 4]
```diff
@@ -13,6 +13,7 @@
 #include "ngraph/util.hpp"
 #include "ngraph/validation_util.hpp"
 #include "ngraph/opsets/opset1.hpp"
+#include "ngraph_ops/type_relaxed.hpp"
```
Comment (eshoguli) [inline inference-engine/src/transformations/src/ngraph_ops/deconvolution_ie.cpp:16]: we don't need it

File: inference-engine/src/low_precision_transformations/src/transformer.cpp [hunk 5]
```diff
@@ -327,6 +329,7 @@ TypeRelaxedReplacer::TypeRelaxedReplacer() {
     make_matcher_type_relaxed<opset1::Clamp>(this);
     make_matcher_type_relaxed<opset1::Concat>(this);
     make_matcher_type_relaxed<opset1::Convolution>(this);
+    make_matcher_type_relaxed<opset1::ConvolutionBackpropData>(this);
```
Comment (eshoguli) [inline inference-engine/src/low_precision_transformations/src/transformer.cpp:332]: Use `NGRAPH_RTTI_DECLARATION` for runtime type info declaration of type relaxed `Deconvolution` operation, please. See `Convolution` as example.

**Review by eshoguli — COMMENTED** [review]
Partially was done. Tests are not reviewed - WIP.

File: inference-engine/src/low_precision_transformations/src/network_helper.cpp [hunk 6]
```diff
@@ -1525,6 +1549,50 @@ std::shared_ptr<Node> NetworkHelper::toScalarIfPossible(std::shared_ptr<Node> no
...
+        const auto intType = intNode->get_input_element_type(0);
+        if (intType == element::u8 || intType == element::i8) {
+            min = DataPrecision::getMinValue(intType, 256) - 0.5f;
+            max = DataPrecision::getMaxValue(intType, 256) + 0.5f;
+        } else {
+            return false;
+        }
+        auto subtractConst = as_type_ptr<opset1::Constant>(node->get_input_node_shared_ptr(1));
+        if (!subtractConst) {
+            subtractConst = as_type_ptr<opset1::Constant>(node->get_input_node_shared_ptr(1)->get_input_node_shared_ptr(0));
```
[the remark this answers is not in the document]
Comment (vzinovie) [reply inference-engine/src/low_precision_transformations/src/network_helper.cpp:1570]: resolved in #4764

File: inference-engine/src/low_precision_transformations/src/network_helper.cpp [hunk 7]
```diff
@@ -1525,6 +1549,50 @@ std::shared_ptr<Node> NetworkHelper::toScalarIfPossible(std::shared_ptr<Node> no
...
+        const auto intNode = is_type<opset1::Convert>(parent) ? parent : node;
+        const auto intType = intNode->get_input_element_type(0);
+        if (intType == element::u8 || intType == element::i8) {
+            min = DataPrecision::getMinValue(intType, 256) - 0.5f;
+            max = DataPrecision::getMaxValue(intType, 256) + 0.5f;
+        } else {
+            return false;
+        }
+        auto subtractConst = as_type_ptr<opset1::Constant>(node->get_input_node_shared_ptr(1));
+        if (!subtractConst) {
```
[the remark this answers is not in the document]
Comment (vzinovie) [reply inference-engine/src/low_precision_transformations/src/network_helper.cpp:1569]: resolved in #4764

File: inference-engine/src/low_precision_transformations/src/network_helper.cpp (hunk 1 above)
Comment (vzinovie) [reply inference-engine/src/low_precision_transformations/src/network_helper.cpp:1587]: resolved in #4764

File: inference-engine/src/low_precision_transformations/src/weightable_layer_transformation.cpp (hunk 2 above)
Comment (vzinovie) [reply inference-engine/src/low_precision_transformations/src/weightable_layer_transformation.cpp]: resolved in #4764

File: inference-engine/src/transformations/include/ngraph_ops/deconvolution_ie.hpp (hunk 3 above)
Comment (vzinovie) [reply inference-engine/src/transformations/include/ngraph_ops/deconvolution_ie.hpp:32]: Don't understand

File: inference-engine/src/transformations/src/ngraph_ops/deconvolution_ie.cpp (hunk 4 above)
Comment (vzinovie) [reply inference-engine/src/transformations/src/ngraph_ops/deconvolution_ie.cpp:16]: Don't understand

File: inference-engine/src/low_precision_transformations/src/transformer.cpp (hunk 5 above)
Comment (vzinovie) [reply inference-engine/src/low_precision_transformations/src/transformer.cpp:332]: Did you mean for `ConvolutionBackpropData`? It has been done already, I suppose.

File: inference-engine/src/low_precision_transformations/src/convolution.cpp [hunk 8]
```diff
@@ -293,6 +282,46 @@ bool ConvolutionTransformation::transform(TransformationContext &context, ngraph
...
+        const auto dataPrecision = getDataPrecisionOnWeights(layer);
+        if ((!supportAsymmetricQuantization) && dataPrecision.hasZeroPoint) {
+            return false;
+        }
+        if (updatePrecisions && !NetworkHelper::checkZeroPoint(fqOnWeights, dataPrecision)) {
+            return false;
+        }
+    } else {
+        return !updatePrecisions || NetworkHelper::checkZeroPoint(dequantization.subtract);
+    }
```
[the remark this answers is not in the document]
Comment (vzinovie) [reply inference-engine/src/low_precision_transformations/src/convolution.cpp:320]: resolved in #4764

File: inference-engine/src/low_precision_transformations/src/convolution_backprop_data.cpp [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+    }
+
+    if (is_type<opset1::Subtract>(onWeights)) {
+        auto& rt = onWeights->get_rt_info();
+        rt["DISABLED_CONSTANT_FOLDING"] = std::make_shared<ngraph::VariantWrapper<std::string>>("");
+    }
+
+    return true;
+}
+bool ConvolutionBackpropDataTransformation::canBeTransformed(const TransformationContext &context,
```
[the remark this answers is not in the document]
Comment (vzinovie) [reply inference-engine/src/low_precision_transformations/src/convolution_backprop_data.cpp:197]: resolved in #4764

**Comment by vzinovie** [comment]
@eshoguli Should we merge #4764 in the first order?

File: inference-engine/src/low_precision_transformations/src/network_helper.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (vzinovie) [reply inference-engine/src/low_precision_transformations/src/network_helper.cpp:1569]: I Suppose, you mean, `return true;`

File: inference-engine/src/transformations/include/ngraph_ops/deconvolution_ie.hpp (hunk 3 above)
Comment (eshoguli) [reply inference-engine/src/transformations/include/ngraph_ops/deconvolution_ie.hpp:32]: I was wrong, was discussed offline - it's not an issue, thanks!

File: inference-engine/src/low_precision_transformations/src/transformer.cpp (hunk 5 above)
Comment (eshoguli) [reply inference-engine/src/low_precision_transformations/src/transformer.cpp:332]: I was wrong, was discussed offline - it's not an issue, thanks!

File: inference-engine/tests/functional/inference_engine/lp_transformations/convolution_backprop_data_transformation.cpp [hunk 10]
```diff
@@ -0,0 +1,314 @@
...
+const std::vector<element::Type> netPrecisions = {
+        element::f32,
+        element::f16
+};
+
+const std::vector<ngraph::Shape> shapes = {
+        ngraph::Shape({ 1, 8, 16, 16 })
+};
+
+const std::vector<ConvolutionBackpropDataTransformationTestValues> testValues = {
```
Comment (v-Golubev) [inline inference-engine/tests/functional/inference_engine/lp_transformations/convolution_backprop_data_transformation.cpp:184]: Do we need to add a test-case with `updatePrecisions=false` here?

File: inference-engine/tests/functional/shared_test_classes/src/base/layer_test_utils.cpp [hunk 11]
```diff
@@ -19,6 +19,7 @@
 #include "ngraph/variant.hpp"
 #include "shared_test_classes/base/layer_test_utils.hpp"
 #include "functional_test_utils/core_config.hpp"
+#include "ngraph/pass/visualize_tree.hpp"
```
Comment (v-Golubev) [inline inference-engine/tests/functional/shared_test_classes/src/base/layer_test_utils.cpp:22]: Please, remove `#include "ngraph/pass/visualize_tree.hpp"`

File: inference-engine/src/low_precision_transformations/src/convolution_backprop_data.cpp [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+                reducedConstant->cast_vector<float>()[0]);
+
+        const auto copyNode = haveOutputShape ?
+            convolutionBackpropData->copy_with_new_inputs({
+                dequantization.multiply->input_value(0),
+                convolutionBackpropData->input_value(1),
+                convolutionBackpropData->input_value(2) }) :
+            convolutionBackpropData->copy_with_new_inputs({
+                dequantization.multiply->input_value(0),
+                convolutionBackpropData->input_value(1) });
```
[the remark this answers is not in the document]
Comment (vzinovie) [reply inference-engine/src/low_precision_transformations/src/convolution_backprop_data.cpp]: Good point! Thank you!

File: inference-engine/tests/functional/inference_engine/lp_transformations/convolution_backprop_data_transformation.cpp (hunk 10 above)
Comment (vzinovie) [reply inference-engine/tests/functional/inference_engine/lp_transformations/convolution_backprop_data_transformation.cpp:184]: Added this case, thanks!

File: inference-engine/tests/functional/shared_test_classes/src/base/layer_test_utils.cpp (hunk 11 above)
Comment (vzinovie) [reply inference-engine/tests/functional/shared_test_classes/src/base/layer_test_utils.cpp]: Removed

**Review by v-Golubev — APPROVED** [review]
Changes look good for me. Thanks!

**Pull request merged**